### ISSUE TITLE
(darcy) Device Does Not Boot After Patching Boot.img

### ISSUE THREAD

**Issue by znbaboy** [topjohnwu/Magisk#8297]
Device: Nvidia Shield TV 2017 (darcy)
Android version: lineage-21.0-20240805-nightly-foster (Android 14)
Magisk version name: Magisk Canary
Magisk version code: 27006

Trying to boot device after patching/flashing boot.img via fastboot/bootloader causes device to return to bootloader interface. Fails to boot.

[Magisk Debug Logs + Patched Boot.img.zip](https://github.com/user-attachments/files/16554169/Magisk.Debug.Logs.%2B.Patched.Boot.img.zip)

**Comment by znbaboy** [comment]
Old issue that was never reopened for some reason: https://github.com/topjohnwu/Magisk/issues/8285

**Comment by hamedsbt** [comment]
Same issue on Nvidia Shield TV 2019 Pro
OS: Lineage-21.0-20240805-nightly-foster (Android 14)
Magisk version name: Magisk Canary
Magisk version code: 27006

**Comment by StupidRepo** [comment]
Was having this issue on `bramble` (Google Pixel 4a 5G) where the boot animation would be really laggy and then after some time of showing the boot animation it would restart. Then all of a sudden it started booting but things like my default home screen app and GBoard configuration was reset? All my apps were still installed.

Also, Magisk appeared to show as if it wasn't installed even though it was.

![Screenshot_20240809-234944_Magisk.png](https://github.com/user-attachments/assets/29b7a334-335c-4622-8a65-04d889f1b0c2)

**Comment by StupidRepo** [comment]
I don't know why that comment was marked off topic as it was very on topic...

**Comment by vvb2060** [comment]
You are using a custom ROM, please ask the ROM builder to contact us.

**Comment by znbaboy** [comment]
> You are using a custom ROM, please ask the ROM builder to contact us. If anyone knows how to fix this, the issue will be reopened.

LineageOS Gitlab issue created: https://gitlab.com/LineageOS/issues/android/-/issues/7545

**Comment by znbaboy** [comment]
Looks like LOS team doesn't want anything to do with it either: https://gitlab.com/LineageOS/issues/android/-/issues/7545

Oh well...

**Comment by hamedsbt** [comment]
> Looks like LOS team doesn't want anything to do with it either: https://gitlab.com/LineageOS/issues/android/-/issues/7545
> 
> Oh well...

I would say, they refusing is reasonable. Unless, we provide an evidence that prove the bug is related to LOS.

**Comment by znbaboy** [comment]
> > Looks like LOS team doesn't want anything to do with it either: https://gitlab.com/LineageOS/issues/android/-/issues/7545
> > Oh well...
> 
> I would say, their refusing is reasonable. Unless, we provide an evidence that prove the bug is related to LOS.

I agree. Unfortunately, we are at the mercy of the developers.

**Comment by znbaboy** [comment]
> You are using a custom ROM, please ask the ROM builder to contact us. If anyone knows how to fix this, the issue will be reopened.

LineageOS is does not support root so we cannot get any support on their end. Is there any reason you believe it is an issue with the ROM? Something I can provide the devs there at-least? Right now we are at a dead end...

**Comment by vvb2060** [comment]
Without help from the ROM builder, troubleshooting the issue could be challenging. Can you get kmsg?

**Comment by hamedsbt** [comment]
@vvb2060 Hi, thank you for your support.
How can I get the kmsg log from my device? I've read https://www.xda-developers.com/how-to-take-logs-android and it suggests that root access is required to get the logs from /sys/fs/pstore/console-ramoops or using the command adb shell su -c dmesg > dmesg.log. However, I don't have root access!

**Comment by hamedsbt** [comment]
@vvb2060 Hi, thank you for your support.

This is Unsuccessful boot /sys/fs/pstore/console-ramoops file after install Magisk Canary v27006 
[console-ramoops-0-magisk.zip](https://github.com/user-attachments/files/16647879/console-ramoops-0.zip)

And this is success boot /sys/fs/pstore/console-ramoops file without Magisk:
[console-ramoops-0.zip](https://github.com/user-attachments/files/16647881/console-ramoops-0.zip)

I'm ready to be a volunteer to install any modified Magisk file(s) to overcome this issue.

Thank you!

**Comment by vvb2060** [comment]
Don't use canary, install debug version and upload kmsg.

**Comment by hamedsbt** [comment]
@vvb2060
I've installed [Magisk app-debug.apk (canary-27006)](https://github.com/topjohnwu/Magisk/releases/download/canary-27006/app-debug.apk), but after that, I couldn't boot Android successfully. So, I went to recovery and pulled the /sys/fs/pstore/console-ramoops file.
[console-ramoops-0.txt](https://github.com/user-attachments/files/16648802/console-ramoops-0.txt)

Also the result of  ```dmesg > /sdcard/dmesg.log``` is:
[dmesg.log](https://github.com/user-attachments/files/16648826/dmesg.log)


Are these okey?

**Comment by vvb2060** [comment]
> For Android 10 devices, the bootloader MUST NOT pass skip_initramfs to the kernel command-line. 

https://source.android.com/docs/core/ota/dynamic_partitions/implement#system-as-root-changes

**Comment by hamedsbt** [comment]
> > For Android 10 devices, the bootloader MUST NOT pass skip_initramfs to the kernel command-line.
> 
> https://source.android.com/docs/core/ota/dynamic_partitions/implement#system-as-root-changes

I'm not a kernel expert. What can I do? The ROM maintainer doesn't answer any questions about Magisk.

**Comment by hamedsbt** [comment]
@vvb2060 I unpacked the boot.img file (```magiskboot -h boot.img```) I want to apply the two kernel parameters you mentioned to the cmdline of the header file. Could you confirm if the format is correct.  
```cmdline=androidboot.boot_devices=sdhci-tegra.0,tegra-sata.0,sdhci-tegra.3 skip_initramfs=0 androidboot.force_normal_boot=1```

Thank you.

**Comment by vvb2060** [comment]
https://github.com/LineageOS/android_kernel_samsung_sdm429/commit/a3e0f795018bfae4c3e805c1418a39a6611d6346

**Comment by hamedsbt** [comment]
@vvb2060 Thank you for your reply.
I can't understand C or C++ code clearly. Could you confirm if the format is correct please?
```cmdline=androidboot.boot_devices=sdhci-tegra.0,tegra-sata.0,sdhci-tegra.3 skip_initramf=0 want_initramf=1 androidboot.force_normal_boot=1```

Thank you.

**Comment by hamedsbt** [comment]
@aviraxp Thank you for your help but I couldn't understand what you mean.

**Comment by vvb2060** [comment]
try https://install.appcenter.ms/users/vvb2060/apps/magisk/distribution_groups/public

**Comment by curiouslux408** [comment]
Device: Xiaomi Xiaomi 11T is agate
Android version: V14.0.5.0.TKWTWXM
Magisk version name: Magisk Canary
Magisk version code: 27007

Trying to boot device after patching/flashing boot.img via fastboot/bootloader causes device to return to bootloader interface. Fails to boot.

**Comment by hamedsbt** [comment]
> try https://install.appcenter.ms/users/vvb2060/apps/magisk/distribution_groups/public

Hi, It's working. Thank you! I also attached the  ```console-ramoops-0``` file:
[console-ramoops-0](https://github.com/user-attachments/files/16747033/console-ramoops-0.txt)

**Comment by hamedsbt** [comment]
I installed this version, but the device doesn't boot again!
![Screenshot_20240908-221419_Trebuchet](https://github.com/user-attachments/assets/fd428fa1-f899-420d-ab52-c7f224c51c97)

**Comment by aviraxp** [comment]
The fix has to be reverted because it breaks more legacy sar + 2si devices. Pleaseet maintainer pick https://github.com/LineageOS/android_kernel_lge_sdm845/commit/026011c9e651996d73a201e47770c425b0c55700 into kernel instead.

**Comment by hamedsbt** [comment]
But unfortunately the maintainer said that he wasn't interested in applying any changes to support Magisk!

**Comment by hamedsbt** [comment]
@vvb2060 Do you update your version recently? Can I update magisk with your release? Thank you.

**Comment by znbaboy** [comment]
sad to see this is still an issue. Just tested on the latest LOS build

**Comment by znbaboy** [comment]
this issue still persists on lineage-22.2-20250512-nightly-foster-signed.zip & Magisk 29.0

**Comment by znbaboy** [comment]
sadly, issue is still present using **Magisk 30.6** on **lineage-22.2-20251208-nightly-foster**

**Comment by RunoHawk** [comment]
Hello, I've had the same issue of device not booting after flashing magisk boot image.
However I found, a bit less than optimal, workaround (as for every update I have to repeat the whole thing)

Last version I used was lineage-23.0-20251220-nightly-rhode for motorola g52
1. Updated system via build in manager (settings -> system -> update)
2.  I followed instruction from "Upgrade to a higher version of LineageOS (e.g. lineage-21 -> lineage-23.0)" (in my case https://wiki.lineageos.org/devices/rhode/upgrade/), basically sideloading the same OS version and gapps, all files remain on device, but please read the steps carefully
3. Once booted I remove both magisk apps (normal and the "private space" copy)
4. Reinstall the app, patch boot.img and flash it in bootloader
5. System boots correctly, magisk is installed and operational

I haven't tried updating system directly via adb yet, will try when new version come out but should work just fine too
Hopefully it's gonna work on other devices too

**Comment by gtorresini** [comment]
Device: NVIDIA Shield TV 2017 (P2897), codename foster/darcy
ROM: LineageOS 22.2 (Android 15), lineage-22.2-20260720-nightly-foster-signed, build fingerprint 5267db34f0
Bootloader: unlocked (fastboot getvar unlocked → yes), secure: No, no vbmeta partition in the partition table at all (bootloader, recovery, boot, dtb, system, cache, userdata, rp4 only)

Symptom: every Magisk-patched boot.img I've tried fails to boot — device returns straight to the fastboot/bootloader interface, no crash log, no display output beyond a blank screen. Same symptom as #1750, #2063, #8285, #8297.

Versions tested, all against this exact build, all with the same result:
- v30.7, v27.0 — patched via boot_patch.sh run unprivileged through adb shell (binaries extracted directly from the release APK), flashed with fastboot flash boot
- v25.2, v23.0 — patched via TWRP's twrp install (genuine root context inside recovery), flashed live to the boot partition
- v20.2 — installer fails before even reaching the patch step (Cannot mount /system — this version predates System-as-Root and doesn't understand this device's current partition layout, so it's untestable on current LineageOS, not a data point either way)

All four working-installer versions (v23.0 through v30.7) bootloop identically, regardless of whether the patch was applied unprivileged via boot_patch.sh or with real root via TWRP.

Ruled out directly on-device (not inferred):
- Wrong source boot.img — sha256sum matches the one embedded in the flashed zip exactly
- System silently updated to a different build than the patched kernel — ro.build.date still matches the flashed zip exactly, no drift
- AVB/vbmeta rejection — no vbmeta partition exists on this device; Magisk's own installer detects this and patches vbmeta flags into the boot image directly (No vbmeta partition, patch vbmeta in boot image), tested and still fails
- Bootloader signature enforcement — secure: No
- Header/ramdisk corruption from magiskboot repack — binary boot image header diff shows only the expected ramdisk-size growth; kernel and cmdline are byte-for-byte untouched
- Stale BCB/misc-partition state carried over from an earlier failed attempt — a RAM-only fastboot boot of the patched image (nothing persisted to any partition) reproduces the identical bootloop, so it isn't a leftover flag

**Comment by gtorresini** [comment]
Solved on my device. @vvb2060 already identified the mechanism back in #15 — what was missing is why the cmdline can't be fixed on this hardware, and what the kernel is actually doing with the flag.

**The cmdline, read off the device (LOS 22.2, darcy, rooted so I could finally read it — `/proc/cmdline` is shell-unreadable on this build):**

```
tegraid=21.1.2.0.0 ... androidboot.bootloader=32.00.2019.50-t210-030626ef skip_initramfs ro root=/dev/mmcblk0p19 rootwait init=/init rootdelay=1 androidboot.verifiedbootstate=orange ... androidboot.boot_devices=sdhci-tegra.0,tegra-sata.0,sdhci-tegra.3
```

`skip_initramfs` is there, bare and valueless, straight from cboot. No `androidboot.force_normal_boot` anywhere in the line, and this device has no `/proc/bootconfig` at all. Same config @hamedsbt's 2024 log showed.

**Why the idea in #16 can't work.** magiskinit evaluates in this order (v30.7, `native/src/init/init.rs`):

```
skip_initramfs      -> legacy_system_as_root()
force_normal_boot   -> first_stage()
/sbin/recovery ...  -> recovery_or_charger()
check_two_stage()   -> first_stage()
```

`skip_initramfs` is checked first and short-circuits the rest, so adding `androidboot.force_normal_boot=1` to the boot.img header cmdline never gets looked at. cboot appends its own args to the header cmdline too, so the header can only add tokens, never remove them. That question went unanswered in the thread — the answer is no.

The device is genuinely 2SI: both `/second_stage_resources` and `/system/bin/init` exist in the stock ramdisk, so `check_two_stage()` would return true and route to `first_stage()` if it ever got that far.

**The part I haven't seen mentioned.** The LineageOS kernel already carries half of the commit @aviraxp linked in #19/#25. Strings in the stock 4.9 kernel:

```
Diag kernel: ignore skip_initramfs to disable system as root
```

So the `init/initramfs.c` + `usr/Kconfig` half of `026011c` is already in. What's missing is the `fs/proc/cmdline.c` half, which rewrites `skip_initramf` -> `want_initramf` in `/proc/cmdline`. The kernel ignores the flag for its own boot and then hands it to magiskinit unchanged. That's the entire bug — magiskinit believes something the kernel itself already decided was meaningless.

Which means for the maintainer this isn't "adding Magisk support" (re #26). It's finishing a patch that's already half shipped.

**Workaround, no kernel rebuild.** Rename the comparison literal inside magiskinit so it stops matching, after Magisk has patched the image. 4 bytes; the kernel is left byte-identical:

```
magiskboot unpack magisk_patched.img
magiskboot cpio ramdisk.cpio "extract init init.bin"
magiskboot hexpatch init.bin 736B69705F696E697472616D667300 77616E745F696E697472616D667300
magiskboot cpio ramdisk.cpio "add 0750 init init.bin"
magiskboot repack magisk_patched.img fixed.img
```

With `skip_initramfs` no longer matching, it falls through to `check_two_stage()` -> `first_stage()`, which is the correct path here.

Result: boots normally, magiskd running, `uid=0(root) gid=0(root) groups=0(root) context=u:r:magisk:s0`. Flashed to boot, survives reboots. Magisk 30.7 on `lineage-22.2-20260720-nightly-foster-signed`.

This is effectively what vvb2060's AppCenter build did in #21 (confirmed working by @hamedsbt in #23) before it was reverted in #25 for breaking legacy SAR + 2SI devices. Reverting it upstream was the right call — the behaviour only makes sense per-device, which is why doing it in the image seems like the right place for it until the kernel gets the other half of `026011c`.

**Attached** (`.tar.gz`, GitHub won't take `.img`/`.sh` directly): a script that does the above, my patched boot.img, sha256sums, and a README with step-by-step instructions — getting `magiskboot`, pulling the right stock boot.img, patching it either through the app or `boot_patch.sh` over adb, then testing and flashing. The script works against any Magisk version since it only touches the ramdisk, and fails loudly if the string isn't found. Usage is just:

```
./patch-magiskinit-skip-initramfs.sh magisk_patched.img fixed.img
fastboot boot fixed.img        # test in RAM first, nothing is written
fastboot flash boot fixed.img  # only if that worked
```

[magisk-skip_initramfs-workaround-foster.tar.gz](https://github.com/user-attachments/files/30381932/magisk-skip_initramfs-workaround-foster.tar.gz)

---

*Disclosure: I worked through this with Claude (Anthropic's Claude Code). The cmdline dumps, kernel strings, checksums and boot tests are all from my own device; the Magisk/kernel source reading and the script came out of that session.*

**Comment by hamedsbt** [comment]
It seems that we, as NVIDIA Shield owners, may have to maintain a **foster-specific Magisk fork**.

The required changes are small and should ideally be considered for mainstream Magisk, but they cannot safely be enabled globally because `skip_initramfs` has legitimate meaning on other devices.

On foster/darcy, the situation is different:

* cboot always supplies `skip_initramfs`.
* The LineageOS kernel already ignores `skip_initramfs` for its own system-as-root decision.
* However, the same flag remains visible in `/proc/cmdline`.
* `magiskinit` sees it first and assumes the legacy SAR path, before it gets a chance to detect that the device is actually 2SI.
* Adding `androidboot.force_normal_boot=1` does not solve this because `skip_initramfs` is evaluated first, and cboot appends its own arguments anyway.

The current workaround is to patch the `skip_initramfs` string in `magiskinit` so that Magisk no longer takes that path. This works, but the modification is lost whenever the boot image is replaced by a ROM update and has to be applied again.

A foster-specific Magisk build would be much cleaner: the required change could be applied automatically whenever Magisk patches the boot image, eliminating the need for users to manually hex-patch `magiskinit` after every LineageOS update.

This does **not** mean Magisk should simply ignore `skip_initramfs` on every device. That would risk breaking devices where the flag is still meaningful. The behavior needs to be conditional on the device/boot configuration.

Ideally, the upstream solution would be even better: fix the LineageOS/Tegra kernel so that the `skip_initramfs` argument that the kernel itself already ignores is not exposed to `magiskinit` as an active boot-mode signal. Until that is possible, a foster-specific Magisk patch seems to be the most practical way to make rooted LineageOS updates manageable.

So the goal isn't really to maintain a permanently separate Magisk ecosystem. The goal is to identify the **small, device-specific changes required for foster**, get them upstream if possible, and make the workaround automatic in the meantime.

**Comment by hamedsbt** [comment]
I've created a patched version of Magisk v30.7 because I don't have access to my device so it would be great if someone could test it.
Note: You can rename it to .apk and install it (the applied patching changes the signature) or just flash it as a zip file in recovery. 

[Magisk_30.7-foster-patched.apk.zip](https://github.com/user-attachments/files/30649797/Magisk_30.7-foster-patched.apk.zip)
